app/questionnaire/router.py

```diff
@@ -207,12 +205,28 @@ def _get_return_to_location_url(
         location: Location,
         return_to: str,
         return_to_answer_id: Optional[str] = None,
+        return_to_block_id: Optional[str] = None,
     ) -> Optional[str]:
+
+        if not return_to:
+            return None
+
+        if (
+            return_to == "calculated-summary"
+            and return_to_block_id
+            and self._schema.is_block_valid(return_to_block_id)
+        ):
```

Comment:
The checks shouldn't be chained, it has to be nested. if you have calculated summary, but block wasn't valid, we do not need to progress down, we should return None. Same applies to the others, i.e returning to section summary, but section isn't complete, hence we can return None, currently it would then check if we are returning to final summary etc, then eventually return None hence why tests pass.

Reply:
Nested these and had to add some tests to ensure code coverage.